### ISSUE TITLE
Floating IP support for pod and VM (discussion)

### ISSUE THREAD

**Issue by lioncruise** [alauda/kube-ovn#112]
I see kube-ovn project can support pod IP through DHCP and static IP, but we know this pod IP is a virtual intranet IP within the pod IP range (CIDR) that we configured when initializing K8S cluster and its CNI plugin, and the pod IP can only be accessed (ping) by the pod in the same namespace and the host K8S node. Is there any support to assign a static external IP to a pod to associate it with the pod IP, so that external machines can directly communicate (eg. ping ) with the specific pod? (Just like VM's floating IP in OpenStack)

Another assumptions:
1. OVN is originally a VM network plugin to replace OVS and can support floating IP. If the KVM VMs can communicate with each other through cni0 bridge (not through OVS), can we implement floating IP for VM through kube-ovn? 
2. If kube-ovn can support floating IP for pod and VM, can other CNI plugin (eg. Calico, Flannel) implement it too?

**Comment by oilbeater** [comment]
OVN originally support DNAT/SNAT that can provide support for "Floating IP" function. However when comes to Kubernetes/Openstack we still need lots works like schedule pod/vm to a host that the fip resides, automatically failover if node gets down, etc. As there is no native fip concept in Kubernetes, much need to be consider carefully. But for a short answer, kube-ovn do plan to support floating ip :)

For other discussions:
1. kube-ovn need metadata from pod to allocate network. If you want to controller vm and pod together, we recommend schedule the vm in a pod (like kube-virt), so we can reuse the cni protocol.
2. other cni plugin can also implement similar function by iptables nat function or other network tools

**Comment by lioncruise** [comment]
@oilbeater Thanks for your quick reply.

From your reply, kube-ovn don't give the feature of pod's FIP mainly for the instability of pods, rather for the technical reason. I think kube-ovn should support VM network feature for it is totally based on OVN. And I have validate that a native KVM VM can directly use cni0 bridge to communicate with pods and other VMs using cni0 bridge (not the kubevirt's wat that runs a VM in a pod). So is there any support to give these VMs FIP through one CNI plugin (not throught another VM network plugin like OVS, ONV, Neutron)? (All VMs and pods use cni0 bridge) 

I don't want to add other network plugins, or add other vswitch, routers because I think cni0 bridge can support VM and pod network communication well. I think current architecture is a very brief and integrated approach.

Current network architecture:

![image](https://user-images.githubusercontent.com/12690984/65677813-d677ef80-e084-11e9-88f0-c38aca0fb902.png)

**Comment by oilbeater** [comment]
From kube-ovn side, it communicates directly to kubernetes apiserver to get the workload information and decides the network topology.  So it even not cares about whether it's container or vm as the underlay runtime. All the difference has been abstract by CNI spec. 

If the vms are out of Kubernetes control scope, kube-ovn has no way to get required information. If the vms are controlled by Kubernetes and follow the CNI spec, when kube-ovn support fip for pod, vm can also get benefit from it.

**Comment by lioncruise** [comment]
@oilbeater Thanks, I got it. kube-ovn works only if VMs or pods follow CNI spec, because kube-ovn gets the pod or VM's CNI information from Kubernetes.

I add the VM's NIC to cni0 manually, which obviously does not follow the CNI spec... I see. Now I have implemented a CRD+controller to implement the lifecycle management of KVM VM on k8s, but not involving network... Next step, I should consider how to adapt VM to CNI spec and synchronize the related information to k8s if I want to use the OVN-K8S integrated solution.
